pkg/knowledge_base/operational_funcs.go

```diff
@@ -0,0 +1,291 @@
+package knowledgebase
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/klothoplatform/klotho/pkg/construct"
+	"github.com/klothoplatform/klotho/pkg/graph"
+)
+
+type (
+	// ConfigTemplateContext is used to scope the DAG into the template functions
+	ConfigTemplateContext struct {
+		DAG *construct.ResourceGraph
+	}
+
+	// ConfigTemplateData provides the resource or edge to the templates as
+	// `{{ .Self }}` for resources
+	// `{{ .Source }}` and `{{ .Destination }}` for edges
+	ConfigTemplateData struct {
+		Resource construct.ResourceId
+		Edge     graph.Edge[construct.ResourceId]
+	}
+)
+
+func (ctx *ConfigTemplateContext) Parse(tmpl string) (*template.Template, error) {
+	t, err := template.New("config").Funcs(template.FuncMap{
+		"upstream":   ctx.Upstream,
+		"downstream": ctx.Downstream,
+		"fieldValue": ctx.FieldValue,
+		"fieldRef":   ctx.FieldRef,
+
+		"split":                split,
+		"filterMatch":          filterMatch,
+		"mapString":            mapString,
+		"zipToMap":             zipToMap,
+		"keysToMapWithDefault": keysToMapWithDefault,
```

Comment:
nit: quick comment around what each one does? Some are obvious mainly zipToMap is what im unsure about